File: problems-adapt/0001-0100/0094_select-kth-smallest-in-a-search-tree/solution_inorder_iterative.ts
function selectKthSmallest(root: TreeNode | null, k: number): number {
    // In-order traversal of a BST visits values in ascending order, so the
    // kth visit is the kth smallest. The explicit stack simulates the
    // recursion, keeping space proportional to the tree height.
    const stack: TreeNode[] = [];
    let node: TreeNode | null = root;
    while (node || stack.length) {
        // Push and descend the left spine as far as possible.
        while (node) {
            stack.push(node);
            node = node.left;
        }
        // Left spine exhausted: popping is the "visit".
        node = stack.pop()!;
        k -= 1;
        // Early stop: the unvisited remainder is never touched.
        if (k === 0) {
            return node.val;
        }
        node = node.right;
    }
    return -1;
}

File: problems/0201-0300/0230_select-kth-smallest-in-a-search-tree/solution_inorder_iterative.py
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def selectKthSmallest(self, root: TreeNode | None, k: int) -> int:
        # In-order traversal of a BST visits values in ascending order, so
        # the kth visit is the kth smallest. The explicit stack simulates the
        # recursion, keeping space proportional to the tree height.
        stack = []
        node = root
        while node or stack:
            # Push and descend the left spine as far as possible.
            while node:
                stack.append(node)
                node = node.left
            # Left spine exhausted: popping is the "visit".
            node = stack.pop()
            k -= 1
            # Early stop: the unvisited remainder is never touched.
            if k == 0:
                return node.val
            node = node.right
        return -1
